### ISSUE TITLE
NewType support with Pydantic

### ISSUE THREAD

**Issue by tungalbert99** [brentyi/tyro#105]
I'm curious how NewTypes are handled underneath the hood and if there's support for it or if I have to modify the type hint. I have something like

`Microliter = NewType("Microliter", float)` which helps keep track of units and in a pydantic model it ends up resolving okay, but not in tyro

```
class Sample(BaseModel) 
     initial_volume: Microliter = pydantic.Field(50) 
```

Underneath, tyro hits a signature error with `typing.Microliter is not supported by signature` 

```
  File ".../tyro/_fields.py", line 448, in _try_field_list_from_callable
    return _try_field_list_from_general_callable(f, cls, default_instance)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../tyro/_fields.py", line 852, in _try_field_list_from_general_callable
    params = list(inspect.signature(f).parameters.values())
```

**Comment by tungalbert99** [comment]
I think `Annotated[]` is a decent work around for now which is non-breaking.

**Comment by brentyi** [comment]
Hi Albert, thanks for filing this issue!

Adding support for `NewType` annotations has been on my mind for a while. It also isn't very hard; I just submitted a PR.

Here's an example script that I added to the tests, which should work after I #106 is merged + I think a bit harder about whether there are edge cases:

```python
from typing import NewType, Tuple

import pydantic
from pydantic import BaseModel
from typing_extensions import Annotated

import tyro

Microliter = NewType("Microliter", int)


class Measurements(BaseModel):
    single: Microliter = pydantic.Field(10)
    renamed_single: Annotated[
        Microliter, tyro.conf.arg(name="other_single")
    ] = pydantic.Field(10)
    pair: Tuple[Microliter, Microliter] = pydantic.Field((20, 30))


IncorrectMeasurements = NewType("IncorrectMeasurements", Measurements)

tyro.cli(IncorrectMeasurements)
```

![image](https://github.com/brentyi/tyro/assets/6992947/0dfae7a5-c776-4eff-b4d9-375a23f7b40c)

**Comment by tungalbert99** [comment]
Awesome thank you! Also loved the talk you gave to our company last year :). I think your library has the most comprehensive features out there and looks like it's going well!

**Comment by brentyi** [comment]
Okay, sounds good! I'll do a release within the next few days, but if you want a hack in the meantime you can also do:
```python
from typing import TYPE_CHECKING, NewType

if TYPE_CHECKING:
    Microliter = NewType("Microliter", int)
else:
    Microliter = int
```

**Comment by brentyi** [comment]
Okay, I just did a release and this should be fixed in `0.6.1`!
